### ISSUE TITLE
File loading speeds

### ISSUE THREAD

**Issue by Jon0** [SFTtech/openage#327]
There is an issue with the `recurse_data_files` function : https://github.com/SFTtech/openage/blob/master/libopenage/util/file.h#L85 and its use in loading the game data files.

Sometimes the loading works in a second, but usually this function seems stuck or takes a long time (10-30 mins)

I tried Monitoring the rate files get loaded, they either get loaded very fast (1000+ game data files /sec)
or very slowly (~50 files /sec), Not sure if this is my file system, or something in this project. 

Any ideas about fixing this?
##

**Comment by mic-e** [comment]
Have you tried putting your assets into a `tmpfs` to rule out any local disk/file system issues?

**Comment by mic-e** [comment]
It would be nice to have some profiling information on this; how about using the gperftools-profiler code that's already integrated (put calls to `external_profiler.start()`, `.stop()` and `.show_results()` before and after the relevant method call. The output of `time (1)` may also yield insights (it tells you whether the time is spent doing I/O in the kernel, or work in the process).

**Comment by mic-e** [comment]
At first glance, 50 files / second seems like the rate that you'd expect if the HDD has to seek twice per file (once for file metadata, once for the file?), e.g. because your system does not do any caching, for some reason.

**Comment by Jon0** [comment]
The slow rate leads to a total 10+ minute loading time. 

Most often the fast loading time occurs straight after booting, and running openage.

**Comment by Jon0** [comment]
running ```free``` after the loading:
```
             total       used       free     shared    buffers     cached
Mem:       4029420    3877424     151996      46004      34736     692048
-/+ buffers/cache:    3150640     878780
Swap:            0          0          0
```

**Comment by TheJJ** [comment]
related: #70 #124 
but the real solution is using nyan and implementing it efficiently. preferrably in a way not everything hast to be loaded at game start, but just the relevant things (e.g. the civ selection doesn't need the available terrains).
And for detecting errors in the files we'd have a separate integrity checker.

**Comment by Jon0** [comment]
It is most likely the large number of small files which causes the 20mb gamedata directory to take far longer loading than the 200mb directory of graphic assets...

**Comment by mic-e** [comment]
Experiments have revealed that openage isn't doing anything wrong _per se_. Reading the `assets/converted/gamedata` folder via `tar c` or `cp -r` takes just about as long as openage takes. The issue seems to be a Kernel/HDD caching issue of an unspecified nature.

This is thus not a bug in openage, especially not a bug in the C++ part of openage. Instead it's a design issue related to limitations of legacy hardware (high-latency rotating disks) and maybe a mis-configuration issue of the Kernel (I still am not sure what exactly causes this).

I have contributed a commit to avoid the call to `stat (2)` to @Jon0's current PR, which I hope will slightly improve loading times.

However, as an intermediate solution that certainly will push loading times to a below-1-second level, you should make `assets/converted/gamedata` a `tmpfs` (or buy an SSD):

    cd assets/converted
    mv gamedata gamedata-disk
    mkdir gamedata
    mount -t tmpfs tmpfs gamedata
    cp -r gamedata-disk/* gamedata

(You'll have to run the last two commands after every reboot).

**Comment by TheJJ** [comment]
Hopefully fixed by #86 and #28...

**Comment by TheJJ** [comment]
This should was more or less fixed by #685 :)
